File: mobile/src/api/qbank.ts
import { client } from './client';

// ─── Types ────────────────────────────────────────────────────────────────────

export interface Question {
  id: string;
  test: string;
  subject: string;
  subtopic: string | null;
  stem: string;
  options: Record<string, string>;   // {"A": "...", "B": "...", ...}
  status: string;
}

export interface TodayQueueResponse {
  questions: Question[];
  srs_due_count: number;
}

export interface AttemptResult {
  is_correct: boolean;
  correct_key: string;
  explanation: string | null;
}

export interface MasteryItem {
  subject: string;
  test: string;
  attempts: number;
  correct: number;
  rate: number;    // 0.0 – 1.0
}

export interface QBankExamSession {
  id: number;
  test_type: string;
  question_ids: string[];
  answers: Record<string, string> | null;
  started_at: string;
  submitted_at: string | null;
  score_pct: number | null;
  by_subject: Record<string, { correct: number; total: number }> | null;
}

export interface QBankExamResult {
  id: number;
  score_pct: number;
  correct: number;
  total: number;
  by_subject: Record<string, { correct: number; total: number }>;
}

// ─── API calls ────────────────────────────────────────────────────────────────

export const getTodayQuestions = async (): Promise<TodayQueueResponse> => {
  const res = await client.get('/students/me/qbank/today');
  return res.data;
};

export const getQuestion = async (id: string): Promise<Question> => {
  const res = await client.get(`/students/me/qbank/questions/${id}`);
  return res.data;
};

export const recordAttempt = async (
  question_id: string,
  selected_key: string,
): Promise<AttemptResult> => {
  const res = await client.post('/students/me/qbank/attempts', {
    question_id,
    selected_key,
  });
  return res.data;
};

export const getMastery = async (): Promise<MasteryItem[]> => {
  const res = await client.get('/students/me/qbank/mastery');
  return res.data;
};

export const startQBankExam = async (test_type: 'temel' | 'klinik'): Promise<QBankExamSession> => {
  const res = await client.post('/students/me/qbank/exams', { test_type });
  return res.data;
};

export const submitQBankExam = async (
  session_id: number,
  answers: Record<string, string>,
): Promise<QBankExamResult> => {
  const res = await client.post(`/students/me/qbank/exams/${session_id}/submit`, { answers });
  return res.data;
};

export const getQBankExam = async (session_id: number): Promise<QBankExamSession> => {
  const res = await client.get(`/students/me/qbank/exams/${session_id}`);
  return res.data;
};

// ─── Drill Mode ───────────────────────────────────────────────────────────────

export interface SubjectItem {
  subject: string;
  subtopics: string[];
}

export interface DrillQueueResponse {
  questions: Question[];
  subject: string;
  subtopic: string | null;
}

export const getSubjects = async (): Promise<SubjectItem[]> => {
  const res = await client.get('/students/me/qbank/subjects');
  return res.data;
};

export const getDrillQuestions = async (
  subject: string,
  subtopic?: string,
  limit = 10,
): Promise<DrillQueueResponse> => {
  const params: Record<string, string | number> = { subject, limit };
  if (subtopic) params.subtopic = subtopic;
  const res = await client.get('/students/me/qbank/drill', { params });
  return res.data;
};

// ─── Subtopic Mastery ─────────────────────────────────────────────────────────

export interface SubtopicMasteryItem {
  subtopic: string;
  attempts: number;
  correct: number;
  rate: number;
}

export const getSubtopicMastery = async (subject: string): Promise<SubtopicMasteryItem[]> => {
  const res = await client.get(`/students/me/qbank/mastery/${encodeURIComponent(subject)}`);
  return res.data;
};

File: backend/app/api/v1/qbank.py
"""
QBank API routes
"""
import uuid
from typing import List
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from sqlalchemy import select

from app.core.db import get_db
from app.models import User
from app.models.qbank import Question, QBankExamSession
from app.api.deps import get_current_user
from app.schemas.qbank import (
    QuestionOut,
    AttemptCreate,
    AttemptResult,
    TodayQueueOut,
    MasteryOut,
    QBankExamCreate,
    QBankExamOut,
    QBankExamSubmitIn,
    QBankExamResult,
    SubjectListOut,
    DrillQueueOut,
    SubtopicMasteryOut,
)
from app.services import qbank_service

router = APIRouter(prefix="/students/me/qbank", tags=["qbank"])


@router.get("/today", response_model=TodayQueueOut)
def get_today_queue(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Return today's question queue: SRS reviews + adaptive new questions (up to 10 total)."""
    questions, srs_due_count = qbank_service.build_today_queue(db, current_user.id)
    return TodayQueueOut(questions=questions, srs_due_count=srs_due_count)


@router.get("/questions/{question_id}", response_model=QuestionOut)
def get_question(
    question_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Return a single approved question (correct_key and explanation withheld)."""
    question = db.get(Question, question_id)
    if not question or question.status == "retired":
        raise HTTPException(status_code=404, detail="Question not found")
    return question


@router.post("/attempts", response_model=AttemptResult)
def record_attempt(
    data: AttemptCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Record an answer, update SRS state, return correct key and explanation."""
    try:
        return qbank_service.record_attempt(
            db, current_user.id, data.question_id, data.selected_key
        )
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc))


@router.get("/subjects", response_model=List[SubjectListOut])
def get_subjects(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Return distinct subjects with subtopics from approved questions (for drill selector)."""
    return qbank_service.get_subjects_list(db)


@router.get("/drill", response_model=DrillQueueOut)
def get_drill_queue(
    subject: str = Query(..., description="Subject to drill"),
    subtopic: str | None = Query(None, description="Optional subtopic filter"),
    limit: int = Query(10, ge=1, le=50),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Return up to `limit` approved questions for subject/subtopic drill.
    Excludes questions answered correctly in the last 7 days.
    """
    questions = qbank_service.get_drill_questions(
        db, current_user.id, subject, subtopic, limit
    )
    return DrillQueueOut(questions=questions, subject=subject, subtopic=subtopic)


@router.get("/mastery", response_model=List[MasteryOut])
def get_mastery(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Return per-subject attempt accuracy for both temel and klinik tracks."""
    return qbank_service.get_mastery(db, current_user.id)


@router.get("/mastery/{subject}", response_model=List[SubtopicMasteryOut])
def get_subtopic_mastery(
    subject: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Return per-subtopic attempt accuracy for a given subject."""
    return qbank_service.get_subtopic_mastery(db, current_user.id, subject)


@router.post("/exams", response_model=QBankExamOut)
def start_exam(
    data: QBankExamCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Start a new timed QBank exam session (up to 100 approved questions)."""
    session = qbank_service.start_exam_session(db, current_user.id, data.test_type)
    if not session.question_ids:
        raise HTTPException(
            status_code=422,
            detail=f"No approved questions found for test type '{data.test_type}'",
        )
    return session


@router.get("/exams/{session_id}", response_model=QBankExamOut)
def get_exam_session(
    session_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get an existing exam session."""
    session = db.get(QBankExamSession, session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Exam session not found")
    if session.student_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not your exam session")
    return session


@router.post("/exams/{session_id}/submit", response_model=QBankExamResult)
def submit_exam(
    session_id: int,
    data: QBankExamSubmitIn,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Submit answers and receive scored results. Idempotent — safe to call twice."""
    session = db.get(QBankExamSession, session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Exam session not found")
    if session.student_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not your exam session")

    session = qbank_service.submit_exam_session(db, session, data.answers)
    correct = sum(
        1 for v in (session.by_subject or {}).values() for _ in range(v["correct"])
    )
    # Re-count correctly from by_subject
    correct = sum(v["correct"] for v in (session.by_subject or {}).values())
    total = len(session.question_ids)
    return QBankExamResult(
        id=session.id,
        score_pct=session.score_pct or 0.0,
        correct=correct,
        total=total,
        by_subject=session.by_subject or {},
    )
